runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -10283,49 +10262,108 @@ genHeapAlloc(TR::Node * node, TR::Instruction *& iCursor, bool isVariableLen, TR
+            iCursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRCT, node, sizeReg, fillerLoopLabel);
 
-      if (zeroReg != NULL)
-         {
-         iCursor = generateRRInstruction(cg, TR::InstOpCode::getXORRegOpCode(), node, zeroReg, zeroReg, iCursor);
+            // Zero the remainder.
+            if(isLargeConstantLen)
+               {
+               // The remainder is a constant value between 1 and 256 for large constant length objects and displacement is 0.
+               allocSize = ((allocSize - 1) % 256) + 1;
+               switch (allocSize)
+                  {
+                  case 1:
+                     iCursor = generateSIInstruction(cg, TR::InstOpCode::MVI, node, generateS390MemoryReference(addressReg, 0, cg), 0, iCursor);
+                     break;
+                  case 2:
+                     iCursor = generateSILInstruction(cg, TR::InstOpCode::MVHHI, node, generateS390MemoryReference(addressReg, 0, cg), 0, iCursor);
+                     break;
+                  case 4:
+                     iCursor = generateSILInstruction(cg, TR::InstOpCode::MVHI, node, generateS390MemoryReference(addressReg, 0, cg), 0, iCursor);
+                     break;
+                  case 8:
+                     iCursor = generateSILInstruction(cg, TR::InstOpCode::MVGHI, node, generateS390MemoryReference(addressReg, 0, cg), 0, iCursor);
+                     break;
+                  default:
+                     iCursor = generateSS1Instruction(cg, TR::InstOpCode::XC, node, (allocSize - 1), generateS390MemoryReference(addressReg, 0, cg), generateS390MemoryReference(addressReg, 0, cg), iCursor);
+                     break;
+                  }
+               }
+            else
+               {
+               TR::LabelSymbol *exrlTargetLabel = generateLabelSymbol(cg);
+               iCursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, fillerRemLabel);
+               iCursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, exrlTargetLabel);
+               iCursor = generateSS1Instruction(cg, TR::InstOpCode::XC, node, 0, generateS390MemoryReference(addressReg, 0, cg), generateS390MemoryReference(addressReg, 0, cg), iCursor);
+               iCursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, fillerRemLabel);
+               iCursor = generateRILInstruction(cg, TR::InstOpCode::EXRL, node, tmp, exrlTargetLabel);
+               }
+            srm->reclaimScratchRegister(tmp);
+            srm->reclaimScratchRegister(addressReg);
+            }
+         else
+            {
+            int32_t displacement = 0;
+            while (allocSize > 256)
+               {
+               iCursor = generateSS1Instruction(cg, TR::InstOpCode::XC, node, 255, generateS390MemoryReference(resReg, displacement, cg), generateS390MemoryReference(resReg, displacement, cg), iCursor);
```

Comment:
As I said, `XC + BRCT + ADD` would take around 14 bytes, while 16 XC instruction would take 96 bytes - That is an additional 82 bytes per new array allocation code inlined. I understand desire for removing loops, but without any comparative analysis on the ops level, with just a rational that 16 sequential instructions would be way faster than the loop implementation is not sufficient. I would have expected that the loop implementation would be efficient with hardware aids like branch predictors. 

If you have seen some workloads where we are allocating that many bytes and you do find that just unrolling XC 16 times is making some visible performance improvement, I would like to take a loot at that. Otherwise, Please reduce that unroll limit upto 512 bytes (That will occupy 12 bytes ) compared to 14 bytes with loop,